File: products/yappc/frontend/libs/ai/src/agents/QueryParserAgent.ts
/**
 * Query Parser Agent Implementation
 *
 * Parses natural language queries into structured intents and filters.
 * Handles search, filter, command, question, and navigation intents.
 *
 * @module ai/agents/QueryParserAgent
 * @doc.type class
 * @doc.purpose Natural language query parsing
 * @doc.layer product
 * @doc.pattern AIAgent
 */

import type { AIService } from '../core/AIService';
import { BaseAgent, type ProcessResult } from './BaseAgent';
import type {
    AgentContext,
    ExtractedEntity,
    ItemFilter,
    ParsedQueryAlternative,
    QueryIntent,
    QueryParserInput,
    QueryParserOutput,
} from './types';
import { QueryParserInputSchema, AgentError } from './types';

/**
 * System prompt for query parsing
 */
const QUERY_PARSER_PROMPT = `You are a query parser for a DevSecOps project management platform.
Parse the user's natural language query into a structured format.

Output a JSON object with:
{
  "intent": "search" | "filter" | "command" | "question" | "navigate",
  "entities": [
    {
      "type": "status" | "priority" | "phase" | "owner" | "tag" | "date" | "item_type" | "keyword",
      "value": "extracted value",
      "confidence": 0.0-1.0
    }
  ],
  "filters": {
    "status": ["in-progress", "blocked"],
    "priority": ["high", "critical"],
    "phaseIds": ["development", "security"],
    "tags": ["urgent"],
    "search": "keyword search"
  },
  "confidence": 0.0-1.0
}

Entity mappings:
- Status: not-started, in-progress, blocked, in-review, completed, archived
- Priority: low, medium, high, critical
- Phases: ideation, planning, development, security, testing, deployment, operations
- Item types: feature, story, task, bug, epic

Examples:
- "show me high priority bugs" → intent: search, filters: {priority: ["high"], status: [], search: "bugs"}
- "what's blocking the security phase?" → intent: question, filters: {phaseIds: ["security"], status: ["blocked"]}
- "assign task-123 to John" → intent: command
- "go to settings" → intent: navigate
`;

/**
 * QueryParserAgent for natural language understanding
 */
export class QueryParserAgent extends BaseAgent<
    QueryParserInput,
    QueryParserOutput
> {
    private aiService: AIService;

    constructor(aiService: AIService) {
        super({
            name: 'QueryParserAgent',
            version: '3.0.0',
            description: 'Natural language query parsing for DevSecOps platform',
            capabilities: [
                'intent-detection',
                'entity-extraction',
                'filter-generation',
                'multi-intent',
            ],
            supportedModels: ['gpt-3.5-turbo', 'gpt-4-turbo'],
            latencySLA: 300,
            defaultTimeout: 5000,
        });

        this.aiService = aiService;
    }

    /**
     * Validate input
     */
    protected validateInput(input: QueryParserInput): void {
        const result = QueryParserInputSchema.safeParse(input);
        if (!result.success) {
            throw new AgentError(
                `Invalid input: ${result.error.message}`,
                'VALIDATION_ERROR',
                this.name,
                false
            );
        }
    }

    /**
     * Process query parsing request
     */
    protected async processRequest(
        input: QueryParserInput,
        context: AgentContext
    ): Promise<ProcessResult<QueryParserOutput>> {
        // Try fast local parsing first
        const localResult = this.tryLocalParse(input.query);
        if (localResult && localResult.confidence > 0.9) {
            return {
                data: localResult,
                modelVersion: 'local-v1',
                confidence: localResult.confidence,
            };
        }

        // Fall back to AI parsing for complex queries
        const enrichedPrompt = this.buildParsePrompt(input);

        const response = await this.aiService.complete(enrichedPrompt, {
            model: 'gpt-3.5-turbo',
            temperature: 0.1,
            maxTokens: 500,
            systemPrompt: QUERY_PARSER_PROMPT,
        });

        const parsed = this.parseAIResponse(response.content);

        // Generate suggestions based on parsed result
        const suggestions = this.generateSuggestions(input.query, parsed);
        const alternatives = this.generateAlternatives(input.query, parsed);

        return {
            data: {
                ...parsed,
                suggestions,
                alternatives,
            },
            tokensUsed: response.usage.totalTokens,
            modelVersion: response.model,
            confidence: parsed.confidence,
        };
    }

    /**
     * Try to parse query locally without AI
     */
    private tryLocalParse(query: string): QueryParserOutput | null {
        const lowerQuery = query.toLowerCase().trim();

        // Simple navigation patterns
        const navPatterns: Record<string, QueryParserOutput> = {
            'go to settings': this.createNavResult('settings'),
            'open settings': this.createNavResult('settings'),
            'show dashboard': this.createNavResult('dashboard'),
            'go home': this.createNavResult('home'),
            'open phases': this.createNavResult('phases'),
            'show workflows': this.createNavResult('workflows'),
        };

        if (navPatterns[lowerQuery]) {
            return navPatterns[lowerQuery];
        }

        // Status filter patterns
        const statusMatch = lowerQuery.match(
            /show (?:me |all )?(blocked|completed|in progress|in-progress) (?:items|tasks)?/
        );
        if (statusMatch) {
            const status = statusMatch[1].replace(' ', '-');
            return {
                intent: 'filter',
                entities: [
                    { type: 'status', value: status, confidence: 0.95, startIndex: 0, endIndex: 0 },
                ],
                filters: { status: [status] },
                confidence: 0.95,
                suggestions: [],
                alternatives: [],
            };
        }

        // Priority patterns
        const priorityMatch = lowerQuery.match(
            /(?:show |find )?(high|critical|low|medium) priority (?:items|tasks|bugs)?/
        );
        if (priorityMatch) {
            const priority = priorityMatch[1];
            const type = lowerQuery.includes('bug') ? 'bugs' : undefined;
            return {
                intent: 'search',
                entities: [
                    { type: 'priority', value: priority, confidence: 0.95, startIndex: 0, endIndex: 0 },
                ],
                filters: {
                    priority: [priority],
                    search: type,
                },
                confidence: 0.95,
                suggestions: [],
                alternatives: [],
            };
        }

        return null;
    }

    /**
     * Create navigation result
     */
    private createNavResult(target: string): QueryParserOutput {
        return {
            intent: 'navigate',
            entities: [
                { type: 'keyword', value: target, confidence: 0.98, startIndex: 0, endIndex: 0 },
            ],
            filters: {},
            confidence: 0.98,
            suggestions: [],
            alternatives: [],
        };
    }

    /**
     * Build prompt for AI parsing
     */
    private buildParsePrompt(input: QueryParserInput): string {
        const parts: string[] = [`Query: "${input.query}"`];

        if (input.currentRoute) {
            parts.push(`Current page: ${input.currentRoute}`);
        }

        if (input.persona) {
            parts.push(`User persona: ${input.persona}`);
        }

        if (input.recentQueries?.length) {
            parts.push(`Recent queries: ${input.recentQueries.slice(0, 3).join(', ')}`);
        }

        return parts.join('\n');
    }

    /**
     * Parse AI response into structured output
     */
    private parseAIResponse(content: string): Omit<QueryParserOutput, 'suggestions' | 'alternatives'> {
        try {
            // Extract JSON from response
            const jsonMatch = content.match(/\{[\s\S]*\}/);
            if (!jsonMatch) {
                return this.defaultResponse();
            }

            const parsed = JSON.parse(jsonMatch[0]);

            return {
                intent: this.validateIntent(parsed.intent),
                entities: this.validateEntities(parsed.entities || []),
                filters: this.validateFilters(parsed.filters || {}),
                confidence: typeof parsed.confidence === 'number' ? parsed.confidence : 0.7,
            };
        } catch {
            return this.defaultResponse();
        }
    }

    /**
     * Validate intent
     */
    private validateIntent(intent: unknown): QueryIntent {
        const validIntents: QueryIntent[] = [
            'search',
            'filter',
            'command',
            'question',
            'navigate',
        ];
        if (typeof intent === 'string' && validIntents.includes(intent as QueryIntent)) {
            return intent as QueryIntent;
        }
        return 'search';
    }

    /**
     * Validate entities
     */
    private validateEntities(entities: unknown[]): ExtractedEntity[] {
        if (!Array.isArray(entities)) {
            return [];
        }

        return entities
            .filter((e): e is Record<string, unknown> => typeof e === 'object' && e !== null)
            .map((e) => ({
                type: String(e.type || 'keyword'),
                value: String(e.value || ''),
                confidence: typeof e.confidence === 'number' ? e.confidence : 0.5,
                startIndex: 0,
                endIndex: 0,
            }));
    }

    /**
     * Validate filters
     */
    private validateFilters(filters: unknown): ItemFilter {
        if (typeof filters !== 'object' || filters === null) {
            return {};
        }

        const f = filters as Record<string, unknown>;
        const result: ItemFilter = {};

        if (Array.isArray(f.status)) {
            result.status = f.status.filter((s): s is string => typeof s === 'string');
        }
        if (Array.isArray(f.priority)) {
            result.priority = f.priority.filter((p): p is string => typeof p === 'string');
        }
        if (Array.isArray(f.phaseIds)) {
            result.phaseIds = f.phaseIds.filter((p): p is string => typeof p === 'string');
        }
        if (Array.isArray(f.tags)) {
            result.tags = f.tags.filter((t): t is string => typeof t === 'string');
        }
        if (typeof f.search === 'string') {
            result.search = f.search;
        }

        return result;
    }

    /**
     * Default response for failed parsing
     */
    private defaultResponse(): Omit<QueryParserOutput, 'suggestions' | 'alternatives'> {
        return {
            intent: 'search',
            entities: [],
            filters: {},
            confidence: 0.5,
        };
    }

    /**
     * Generate query suggestions
     */
    private generateSuggestions(
        query: string,
        parsed: Omit<QueryParserOutput, 'suggestions' | 'alternatives'>
    ): string[] {
        const suggestions: string[] = [];

        if (parsed.intent === 'search' && !parsed.filters.status?.length) {
            suggestions.push(`${query} that are blocked`);
            suggestions.push(`${query} in progress`);
        }

        if (parsed.intent === 'filter') {
            suggestions.push('Sort by priority');
            suggestions.push('Group by phase');
        }

        return suggestions.slice(0, 3);
    }

    /**
     * Generate alternative interpretations
     */
    private generateAlternatives(
        _query: string,
        parsed: Omit<QueryParserOutput, 'suggestions' | 'alternatives'>
    ): ParsedQueryAlternative[] {
        const alternatives: ParsedQueryAlternative[] = [];

        // If parsed as search, could also be filter
        if (parsed.intent === 'search' && parsed.filters.search) {
            alternatives.push({
                intent: 'filter',
                entities: parsed.entities,
                confidence: parsed.confidence * 0.8,
            });
        }

        return alternatives;
    }

    /**
     * Check dependencies
     */
    protected async checkDependencies(): Promise<
        Record<string, 'healthy' | 'degraded' | 'unhealthy'>
    > {
        try {
            const aiHealthy = await this.aiService.healthCheck();
            return {
                aiService: aiHealthy ? 'healthy' : 'degraded', // Can fall back to local
                localParser: 'healthy',
            };
        } catch {
            return {
                aiService: 'degraded',
                localParser: 'healthy',
            };
        }
    }
}

File: products/yappc/core/domain/src/main/java/com/ghatana/products/yappc/domain/agent/QueryParserAgent.java
package com.ghatana.products.yappc.domain.agent;

import com.ghatana.ai.llm.ChatMessage;
import com.ghatana.ai.llm.CompletionRequest;
import com.ghatana.ai.llm.LLMGateway;
import com.ghatana.platform.observability.MetricsCollector;
import io.activej.promise.Promise;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Query Parser Agent for natural language search query parsing.
 * <p>
 * Parses natural language queries into structured filters, intents,
 * and actions for the DevSecOps platform.
 * <p>
 * <b>Responsibilities:</b>
 * <ul>
 *   <li>Parse natural language search queries</li>
 *   <li>Extract entities (item IDs, phases, statuses, etc.)</li>
 *   <li>Detect user intent (search, filter, command, question, navigate)</li>
 *   <li>Generate structured filters for querying</li>
 * </ul>
 * <p>
 * <b>Models:</b> Custom NLU + GPT-3.5-turbo for complex queries
 * <p>
 * <b>Latency SLA:</b> 300ms
 *
 * @doc.type class
 * @doc.purpose Natural language query parser agent
 * @doc.layer product
 * @doc.pattern AIAgent
 */
public class QueryParserAgent extends AbstractAIAgent<QueryParserInput, QueryParserOutput> {

    private static final Logger LOG = LoggerFactory.getLogger(QueryParserAgent.class);

    private static final String VERSION = "2.0.0";
    private static final String DESCRIPTION = "Natural language query parsing and intent detection";
    private static final List<String> CAPABILITIES = List.of(
            "intent-detection",
            "entity-extraction",
            "filter-generation",
            "semantic-understanding"
    );
    private static final List<String> SUPPORTED_MODELS = List.of(
            "custom-nlu",
            "gpt-3.5-turbo"
    );

    private static final String PARSER_PROMPT = """
            Parse the following natural language query into a structured format.
            
            Query: "%s"
            Context Route: %s
            Persona: %s
            
            Extract:
            1. Intent: SEARCH | FILTER | COMMAND | QUESTION | NAVIGATE
            2. Entities: item IDs, phases, statuses, priorities, assignees, dates, tags
            3. Filters: Convert entities into filter criteria
            4. Confidence: Your confidence in the parsing (0-1)
            
            Respond in JSON format:
            {
              "intent": "...",
              "entities": [...],
              "filters": {...},
              "action": null or {...},
              "confidence": 0.95
            }
            """;

    // Common patterns for quick parsing
    private static final Pattern ITEM_ID_PATTERN = Pattern.compile(
            "\\b([A-Z]+-\\d+|ITEM-\\d+|TASK-\\d+|BUG-\\d+|VULN-\\d+)\\b",
            Pattern.CASE_INSENSITIVE
    );
    private static final Pattern PRIORITY_PATTERN = Pattern.compile(
            "\\b(high|medium|low|critical|urgent)\\s*priority\\b",
            Pattern.CASE_INSENSITIVE
    );
    private static final Pattern STATUS_PATTERN = Pattern.compile(
            "\\b(open|in[- ]progress|done|completed|blocked|pending|closed)\\b",
            Pattern.CASE_INSENSITIVE
    );
    private static final Pattern PHASE_PATTERN = Pattern.compile(
            "\\b(planning|design|development|testing|security|deployment|monitoring)\\s*phase\\b",
            Pattern.CASE_INSENSITIVE
    );

    private final LLMGateway llmGateway;
    private final boolean useFastPath;

    /**
     * Creates a new QueryParserAgent.
     *
     * @param llmGateway       The LLM gateway for AI completions
     * @param metricsCollector The metrics collector
     * @param useFastPath      Whether to use fast pattern matching before LLM
     */
    public QueryParserAgent(
            @NotNull LLMGateway llmGateway,
            @NotNull MetricsCollector metricsCollector,
            boolean useFastPath
    ) {
        super(
                AgentName.QUERY_PARSER_AGENT,
                VERSION,
                DESCRIPTION,
                CAPABILITIES,
                SUPPORTED_MODELS,
                metricsCollector
        );
        this.llmGateway = llmGateway;
        this.useFastPath = useFastPath;
    }

    /**
     * Creates a QueryParserAgent with fast path enabled.
     */
    public QueryParserAgent(
            @NotNull LLMGateway llmGateway,
            @NotNull MetricsCollector metricsCollector
    ) {
        this(llmGateway, metricsCollector, true);
    }

    @Override
    public void validateInput(@NotNull QueryParserInput input) {
        if (input.query() == null || input.query().isBlank()) {
            throw new IllegalArgumentException("Query cannot be empty");
        }
    }

    @Override
    protected Promise<ProcessResult<QueryParserOutput>> processRequest(
            @NotNull QueryParserInput input,
            @NotNull AIAgentContext context
    ) {
        LOG.debug("Parsing query: {}", input.query());

        // Try fast path first for simple queries
        if (useFastPath) {
            QueryParserOutput fastResult = tryFastParse(input);
            if (fastResult != null && fastResult.parsed().confidence() > 0.8) {
                LOG.debug("Fast path succeeded with confidence: {}", fastResult.parsed().confidence());
                metricsCollector.incrementCounter("agent.query_parser.fast_path", "result", "success");
                return Promise.of(ProcessResult.of(fastResult, fastResult.parsed().confidence()));
            }
        }

        // Fall back to LLM for complex queries
        return parseWithLLM(input, context);
    }

    private QueryParserOutput tryFastParse(QueryParserInput input) {
        String query = input.query().toLowerCase();
        List<QueryParserOutput.ExtractedEntity> entities = new ArrayList<>();
        double confidence = 0.0;

        // Extract item IDs
        Matcher itemMatcher = ITEM_ID_PATTERN.matcher(input.query());
        while (itemMatcher.find()) {
            entities.add(new QueryParserOutput.ExtractedEntity(
                    QueryParserOutput.ExtractedEntity.TYPE_ITEM_ID,
                    itemMatcher.group(),
                    itemMatcher.group().toUpperCase(),
                    itemMatcher.start(),
                    itemMatcher.end(),
                    0.95
            ));
            confidence = Math.max(confidence, 0.9);
        }

        // Extract priorities
        Matcher priorityMatcher = PRIORITY_PATTERN.matcher(query);
        List<String> priorities = new ArrayList<>();
        while (priorityMatcher.find()) {
            String priority = priorityMatcher.group(1).toLowerCase();
            priorities.add(priority);
            entities.add(new QueryParserOutput.ExtractedEntity(
                    QueryParserOutput.ExtractedEntity.TYPE_PRIORITY,
                    priorityMatcher.group(),
                    priority,
                    priorityMatcher.start(),
                    priorityMatcher.end(),
                    0.9
            ));
            confidence = Math.max(confidence, 0.85);
        }

        // Extract statuses
        Matcher statusMatcher = STATUS_PATTERN.matcher(query);
        List<String> statuses = new ArrayList<>();
        while (statusMatcher.find()) {
            String status = statusMatcher.group(1).toLowerCase().replace(" ", "-");
            statuses.add(status);
            entities.add(new QueryParserOutput.ExtractedEntity(
                    QueryParserOutput.ExtractedEntity.TYPE_STATUS,
                    statusMatcher.group(),
                    status,
                    statusMatcher.start(),
                    statusMatcher.end(),
                    0.9
            ));
            confidence = Math.max(confidence, 0.85);
        }

        // Extract phases
        Matcher phaseMatcher = PHASE_PATTERN.matcher(query);
        List<String> phases = new ArrayList<>();
        while (phaseMatcher.find()) {
            String phase = phaseMatcher.group(1).toLowerCase();
            phases.add(phase);
            entities.add(new QueryParserOutput.ExtractedEntity(
                    QueryParserOutput.ExtractedEntity.TYPE_PHASE,
                    phaseMatcher.group(),
                    phase,
                    phaseMatcher.start(),
                    phaseMatcher.end(),
                    0.9
            ));
            confidence = Math.max(confidence, 0.85);
        }

        // If no entities found, return null to trigger LLM parsing
        if (entities.isEmpty()) {
            return null;
        }

        // Determine intent
        QueryParserOutput.ParsedQuery.QueryIntent intent = determineIntent(query, entities);

        // Build filters
        QueryParserOutput.ItemFilter filters = new QueryParserOutput.ItemFilter(
                statuses.isEmpty() ? null : statuses,
                priorities.isEmpty() ? null : priorities,
                phases.isEmpty() ? null : phases,
                null, // assignees
                null, // tags
                null, // searchText
                null, // dateRange
                null  // customFilters
        );

        QueryParserOutput.ParsedQuery parsed = new QueryParserOutput.ParsedQuery(
                intent,
                entities,
                filters,
                null,
                confidence
        );

        return new QueryParserOutput(
                parsed,
                generateSuggestions(input, parsed),
                List.of()
        );
    }

    private QueryParserOutput.ParsedQuery.QueryIntent determineIntent(
            String query,
            List<QueryParserOutput.ExtractedEntity> entities
    ) {
        // Check for navigation keywords
        if (query.contains("go to") || query.contains("navigate") || query.contains("show me")) {
            return QueryParserOutput.ParsedQuery.QueryIntent.NAVIGATE;
        }

        // Check for command keywords
        if (query.contains("create") || query.contains("add") || query.contains("delete") ||
                query.contains("update") || query.contains("assign") || query.contains("move")) {
            return QueryParserOutput.ParsedQuery.QueryIntent.COMMAND;
        }

        // Check for question keywords
        if (query.contains("what") || query.contains("why") || query.contains("how") ||
                query.contains("when") || query.endsWith("?")) {
            return QueryParserOutput.ParsedQuery.QueryIntent.QUESTION;
        }

        // Check for filter keywords
        if (query.contains("filter") || query.contains("only") || query.contains("exclude") ||
                !entities.isEmpty()) {
            return QueryParserOutput.ParsedQuery.QueryIntent.FILTER;
        }

        // Default to search
        return QueryParserOutput.ParsedQuery.QueryIntent.SEARCH;
    }

    private Promise<ProcessResult<QueryParserOutput>> parseWithLLM(
            QueryParserInput input,
            AIAgentContext context
    ) {
        String prompt = String.format(
                PARSER_PROMPT,
                input.query(),
                input.currentRoute() != null ? input.currentRoute() : "unknown",
                input.persona() != null ? input.persona() : "default"
        );

        List<ChatMessage> messages = List.of(
                ChatMessage.system("You are a query parser for a DevSecOps platform. Parse queries into structured filters."),
                ChatMessage.user(prompt)
        );

        CompletionRequest request = CompletionRequest.builder()
                .messages(messages)
                .model("gpt-3.5-turbo")
                .temperature(0.3) // Lower temperature for more consistent parsing
                .maxTokens(512)
                .build();

        return llmGateway.complete(request)
                .map(result -> {
                    // Parse the JSON response
                    QueryParserOutput output = parseJsonResponse(result.getText(), input);
                    return ProcessResult.of(
                            output,
                            result.getTokensUsed(),
                            result.getModelUsed(),
                            output.parsed().confidence()
                    );
                });
    }

    private QueryParserOutput parseJsonResponse(String jsonResponse, QueryParserInput input) {
        // In production, use Jackson/Gson for proper JSON parsing
        // For now, return a default parsed query
        QueryParserOutput.ParsedQuery parsed = new QueryParserOutput.ParsedQuery(
                QueryParserOutput.ParsedQuery.QueryIntent.SEARCH,
                List.of(),
                QueryParserOutput.ItemFilter.empty(),
                null,
                0.75
        );

        return new QueryParserOutput(
                parsed,
                generateSuggestions(input, parsed),
                List.of()
        );
    }

    private List<String> generateSuggestions(QueryParserInput input, QueryParserOutput.ParsedQuery parsed) {
        List<String> suggestions = new ArrayList<>();

        // Add context-based suggestions
        if (parsed.filters().statuses() != null && !parsed.filters().statuses().isEmpty()) {
            suggestions.add("Also show " + getOppositeStatus(parsed.filters().statuses().get(0)));
        }

        if (parsed.filters().priorities() != null && !parsed.filters().priorities().isEmpty()) {
            suggestions.add("Include medium priority items");
        }

        // Add common follow-up queries
        suggestions.add("Sort by due date");
        suggestions.add("Group by assignee");

        return suggestions;
    }

    private String getOppositeStatus(String status) {
        return switch (status.toLowerCase()) {
            case "open" -> "completed items";
            case "completed", "done" -> "open items";
            case "blocked" -> "unblocked items";
            case "in-progress" -> "pending items";
            default -> "other statuses";
        };
    }

    @Override
    protected Promise<Map<String, AgentHealth.DependencyStatus>> doHealthCheck() {
        return llmGateway.complete(
                CompletionRequest.builder()
                        .prompt("test")
                        .maxTokens(1)
                        .build()
        ).map(result -> Map.of(
                "llmGateway", AgentHealth.DependencyStatus.HEALTHY,
                "patternMatcher", AgentHealth.DependencyStatus.HEALTHY
        )).mapException(e -> {
            LOG.error("Health check failed: {}", e.getMessage());
            return new RuntimeException("LLM Gateway unhealthy");
        });
    }
}
